### ISSUE TITLE
Suggestion: Change the default data types of many Nodes to Floats/Ints/Dummy

### ISSUE THREAD

**Issue by ProbablePrime** [Frooxius/NeosPublic#358]
This might be a contentious issue but i wanted to open it.

A lot of the comparison nodes such as:
- \<
- \>
- \<=
- \>=

And I think a lot of the arithmetic nodes:
- \+ 
- \-

Etc

Spawn by default with their inputs set to the double data type(deep blue).

Something I notice a lot of new users do is they have a float or an int from another node and don't realize they can link them together.

Alternatively they pull display ribbons from the inputs and press secondary, this spawns a double input node. 

At this point if they plug in floats or ints they will auto-cast but this opens up added complexities as they then need to understand cast nodes.

To improve the experience here the default data type of these nodes could be changed to a more common data type. 

Another reason is that the input nodes for double are found within the Input->Uncommon folder which is hard for a new user to locate.

Removing this decision point or confusion point from new users would be an awesome move.

**Comment by Frooxius** [comment]
This is a bit tricky, since there's a lot of the nodes, the selection needs to be done in some general way, so I don't have to specify it for each. If there's both int and float, which one should be prioritized? 

It should likely be float since it's more common, but if the user is working with integers that might end up confusing them anyway, at which point they need to understand the datatypes anyway.

**Comment by ProbablePrime** [comment]
I understand the confusion eventually coming but as a counter to that, I never see Double's anywhere so there's added confusion that doubles even show up at all generated from these.

**Comment by shiftyscales** [comment]
Use of the pink dummy type in particular could be really nice as it encourages plugging in varying types and experimenting since it is initially undefined. This also applies to many other operators such as ?: too.

I think the best general option will be to default to dummy where possible as it supports the goal of making it more obvious that the nodes are capable of transforming to accept varying types of inputs without relying on specifying a default for each.

**Comment by Frooxius** [comment]
Yeah float/int is definitely better option there. But I'm just wondering how I should approach it. I think float -> int -> dummy would be best in order of priority based on pragmatic use.

Dummy as default could work too, but I'm afraid it might be more tedious in some cases. I do want to have a system to indicate node overloads in the future which could obviate the issue. Dummy could be used in the meanwhile as default.

**Comment by ProbablePrime** [comment]
I'd agree with Dummy, I love Shift's comment on experimentation. I was teaching two new users that in a world by just slamming random nodes together.

**Comment by mralext20** [comment]
problem with having dummy be the default is now you cant (for example) spawn a + node, and instantly start using it (without spawning atleast one input node to set the type of the +)

**Comment by shiftyscales** [comment]
I specify ?: in particular because of a memory I have associated to it. When I was learning about that node initially, I thought it could only accept a bool input because it defaulted to bool. I just figured that was its inherent function until another user pointed out that you can change its typing by plugging a different type in.

Suddenly the node made way more sense, and its applications became a lot clearer.

**Comment by shiftyscales** [comment]
Following up from the update posted on January 9th, `0.8.19.18996`:
` When spawning new multi-type LogiX nodes, initial variant prefers versions with datatypes in order of dummy -> float -> int -> regular (based on feedback from @ProbablePrime and @Shifty)`

It appears that the internal ordering is incorrect, and prioritizes floats, and ints over dummy.
![2020-01-22 21 09 35](https://user-images.githubusercontent.com/54213390/72958840-49585500-3d5d-11ea-979e-a6443f38bee0.png)

**Comment by Frooxius** [comment]
I've double checked, the order is definitely correct. Problem here is that these nodes don't have an overload with the dummy datatype, so it can't be selected, so the next available type is chosen (float or int).

I could add dummy versions to these, but that's potentially a lot of work involved, since dummy datatype is incapable of these operations, so it would need mockups for them.

**Comment by shiftyscales** [comment]
I see. That does make sense. If it would be a lot of work to implement, it isn't worth implementing at this moment.

The primary purpose of the original suggestion was just as a means to better convey the transformative property of many nodes in their ability to auto-cast to a variety of inputs.

So with this goal in mind- any sort of visual that can help distinguish the types that can be auto-cast would be useful and appreciated.

**Comment by Frooxius** [comment]
I plan on doing that in general, by visually distinguishing types that accept different datatypes. It's also a fair bit of work, but should be better way of handling this, as it will work for all nodes of any type.

**Comment by shiftyscales** [comment]
That sounds like a great option to me. As noted, it's not a priority feature, but would be appreciated in the future when you have the time to implement it. I think this issue could possibly be considered as closed for now, although since he opened it, I think it fair that @ProbablePrime get a say too.

Any subsequent work would be outside of the initial scope of the issue which has been met, and addressed.

**Comment by shiftyscales** [comment]
Thoughts, @ProbablePrime? I can see the merits for both closing, and leaving this particular issue open.

The initial request had been granted, so it could be closed- but Frooxius' proposed solution hasn't been implemented, so it could remain open.

**Comment by JackTheFoxOtter** [comment]
A bit of a note from my point of view as I've stumbled accross a situation in which I assumed a node was accepting a specific input type, only to realize that the node had no overload for that type and the wire wouldn't connect. While exploration and trying to figure out what works and what doesn't can be fun, I would argue in this case it's not a good idea. Using the dummy data type just tells the user that the Node's type hasn't been set yet, but it still doesn't give any information about what overloads are supported.

I think the best solution would be a way to visually indicate what overloads exist for a node. This is a pretty tricky problem, because overloads aren't just a per input socket thing, they define the combination of all socket types, depending on what type would be choosen for socket 1 it could limit what's available for socket 2, and those don't have to be of the same type. Also, since many nodes spawn with a default socket type (keep in mind spawning them with the dummy type to signal that there are overloads isn't necessary if there is another visual representation for overloads), there would probably need to be a way to see overloads even when the types are already set.

I think this will likely tie into the UI rework, and might go hand in hand with providing aliases for node names, and perhaps even including an in-game embedded view of the node's corresponding wiki page. The tricky part of all of these will be to come up with a design that doesn't infringe on LogiX's simple and clean visuals.

And idea that comes to my mind would be to slowly cycle through the colors of each currently possible type per unconnected input socket, respecting the limitations imposed by the types of already connected sockets. If the type of an output socket can vary depending on the input, that should probably default to the dummy color if the output type isn't yet yet clear.
